lib/streamlit/type_util.py

```diff
@@ -712,6 +713,73 @@ def is_pyarrow_version_less_than(v: str) -> bool:
     return version.parse(pa.__version__) < version.parse(v)
 
 
+def _maybe_truncate_table(
+    table: pa.Table, truncated_rows: int | None = None
+) -> pa.Table:
+    """Experimental feature to automatically truncate tables that
+    are larger than the maximum allowed message size. It needs to be enabled
+    via the server.enableArrowTruncation config option.
+
+    Parameters
+    ----------
+    table : pyarrow.Table
+        A table to truncate.
+
+    truncated_rows : int or None
+        The number of rows that have been truncated so far. This is used by
+        the recursion logic to keep track of the total number of truncated
+        rows.
+
+    """
+
+    if config.get_option("server.enableArrowTruncation") is not True:
+        return table
+
+    # This is an optimization problem: We don't know at what row
+    # the perfect cut-off is to comply with the max size. But we want to figure
+    # it out in as few iterations as possible. We almost always will cut out
+    # more than required to keep the iterations low.
+
+    # The maximum size allowed for protobuf messages in bytes:
+    max_message_size = int(config.get_option("server.maxMessageSize")) * int(1e6)
+    # We add 1 MB for other overhead related to the protobuf message.
+    # This is a very conservative estimate, but it should be good enough.
+    table_size = int(table.nbytes) + 1 * int(1e6)
+    table_rows = table.num_rows
+
+    if table_rows > 1 and table_size > max_message_size:
+        # Calculate an approximation of how many rows we need to truncate to.
+        targeted_rows = math.ceil(table_rows * (max_message_size / table_size))
+        # The targeted_rows is just an approximation. We cut out another
+        # 5% of the number of truncated rows to make sure that we don't
+        # execute this truncation logic too many times since it is quite
+        # inefficient and its running in a recursion.
+        rows_overhead = math.floor((table_rows - targeted_rows) * 0.05)
```

Comment:
Hm, I wonder if we should consider calculating the additional rows to truncate in a different way. I'm not sure if this edge case is one that'll actually appear in practice, but theoretically we could be sending dataframes sufficiently large that we truncate, say, 99% of rows, in which case we'll have `rows_overhead` > `targeted_rows` and end up recursing for a long time given that this will cause us to only cut off 5 rows at a time.

Maybe we can just reduce `targeted_rows` by some constant factor <1? Although I'm not sure how effective that'll be in reducing the number of recursions.

Reply:
> in which case we'll have rows_overhead > targeted_rows and end up recursing for a long time given that this will cause us to only cut off 5 rows at a time.

I think if `rows_overhead` > `targeted_rows`, it will actually just directly jump to cutting it down to 1 row -> which will then stop the recursion (because of the `table_rows > 1` check) and return this table with 1 row 

e.g:
```
targeted_rows = 10
rows_overhead = 20
table_rows = 30

max(min(targeted_rows - rows_overhead, table_rows - 10), 1)
max(min(10 - 20, 30 - 10), 1)

-> The inner min will be negative, so it will use 1 as a target because of the outer max. 
```

In theory, this 1 row can still be bigger than the max size. But that is an extreme edge case where it would be fine to just run into the limit exceed error. Usually, you will get at least a couple of 10k rows into 32 MB (the limit in SiS).

Reply:
Overall, I think this code will be able to truncate the table within one or two iterations for the vast majority of cases since, with a couple of 10k rows, we usually should have quite a good statistical average for an average row. 
The main edge cases are with single rows within the table that have a significantly higher bytes footprint.

Reply:
> There's also the option of converting this function to be iterative

I thought about this as well. I have now added an exception for the recursion error. So, in the very edgy edge case where this might happen, it would just ignore truncation and run into the normal max message size exceeding the error.

Reply:
> I think if rows_overhead > targeted_rows, it will actually just directly jump to cutting it down to 1 row -> which will then stop the recursion (because of the table_rows > 1 check) and return this table with 1 row

Oops -- I'm just bad at math 😓